tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+    page object doesn't inherit from something like ``ItemPage[ItemClass]``
+    """
+    item, deps = yield crawl_item_and_deps(StandaloneProduct)
+    assert "UndeclaredProvidedTypeError:" in caplog.text
+    assert item is None
+    assert_deps(deps, {}, size=0)
+
+    # calling the actual page object should still work
+    item, deps = yield crawl_item_and_deps(StandaloneProductPage)
+    assert item == {"name": "standalone product name"}
+    assert_deps(deps, {"page": StandaloneProductPage})
+
+
+@attrs.define
+class Morty:
+    name: str
+
+
+@handle_urls(URL)
+class RickSanchezPage(ItemPage[Morty]):
+    @field
+    def name(self) -> str:
+        return "from basic rick"
+
+
+@handle_urls(URL, instead_of=RickSanchezPage)
+class RickSanchezC137Page(ItemPage[Morty]):
+    @field
+    def name(self) -> str:
+        return "wubba lubba dub dub"
+
+
+@inlineCallbacks
+def test_item_return_with_overrides() -> None:
+    item, deps = yield crawl_item_and_deps(Morty)
+    assert item == Morty(name="wubba lubba dub dub")
+    assert_deps(deps, {"item": RickSanchezC137Page})
+
+    # page from 'instead_of'
+    item, deps = yield crawl_item_and_deps(RickSanchezPage)
+    assert item == Morty(name="wubba lubba dub dub")
+    assert_deps(deps, {"page": RickSanchezC137Page})
+
+    # calling the actual page object should still work
+    item, deps = yield crawl_item_and_deps(RickSanchezC137Page)
+    assert item == Morty(name="wubba lubba dub dub")
+    assert_deps(deps, {"page": RickSanchezC137Page})
+
+
+@attrs.define
+class ProductFromInjectable:
+    name: str
+
+
+@handle_urls(URL, to_return=ProductFromInjectable)
+class ProductFromInjectablePage(Injectable):
+    @field
+    def name(self) -> str:
+        return "product from injectable"
+
+    async def to_item(self) -> ProductFromInjectable:
+        return await item_from_fields(self, item_cls=ProductFromInjectable)
+
+
+@inlineCallbacks
+def test_item_return_from_injectable() -> None:
+    """The case wherein a PageObject inherits directly from ``web_poet.Injectable``
+    should also work, provided that the ``to_item`` method is similar with
+    ``web_poet.ItemPage``:
+    """
+    item, deps = yield crawl_item_and_deps(ProductFromInjectable)
+    assert item == ProductFromInjectable(name="product from injectable")
+    assert_deps(deps, {"item": ProductFromInjectable})
+
+    # calling the actual page object should not work since it's not inheriting
+    # from ``web_poet.ItemPage``.
+    item, deps = yield crawl_item_and_deps(ProductFromInjectablePage)
+    assert item is None
+
+    # However, the page object should still be injected into the callback method.
+    assert_deps(deps, {"item": ProductFromInjectablePage})
+
+
+@handle_urls(URL)
+class PageObjectDependencyPage(ItemPage):
+    async def to_item(self) -> dict:
+        return {"name": "item dependency"}
+
+
+@attrs.define
+class MainProductA:
+    name: str
+    item_from_po_dependency: dict
```

Comment:
I'm not sure we should be supporting dicts as items in injection. There seem to be too many pitfalls.

* ItemPage uses dict by default
* There can be "dict" attributes in general, which are not related at all to items.

So, combination of this could mean that we start to do some weird and unexpected injection. I think we shouldn't inject dicts.

Reply:
I might be missing something but we're currently not supporting `dict`s to be injected.

`item_from_po_dependency` in this case is populated via the page object dependency as:

```python
@field
async def item_from_po_dependency(self) -> dict:
    return await self.injected_page.to_item()
```

It'd be hard for the `ItemProvider` to identify which PO returns the given `dict` item. We only support the following at the moment:

```python
@attrs.define
class MainProductB:                                                                                                                           
    name: str                                                                                                                                 
    item_dependency: ItemDependency
```

Reply:
I misread the code, sorry! Makes sense.